alf/layers.py

```diff
@@ -1092,3 +1092,246 @@ def calc_output_shape(self, input_shape):
+                encoding proposed in [3]. If False, use absolute positional encoding.
+            add_positional_encoding (bool): If True, in addition to use positional
+                encoding for calculating the attention weights, the positional encoding
+                is also concatenated to the attention result so that the attention
+                result can keep the location information better. Note that using
+                this option will increase the number of parameters by about 25%.
+        """
+        super().__init__()
+        if d_k is None:
+            d_k = d_model
+        if d_v is None:
+            d_v = d_model
+        self._q_proj = nn.Parameter(torch.Tensor(d_model, num_heads * d_k))
+        self._k_proj = nn.Parameter(torch.Tensor(d_model, num_heads * d_k))
+        self._v_proj = nn.Parameter(torch.Tensor(d_model, num_heads * d_v))
+        d_a = d_v
+        if add_positional_encoding:
+            d_a = d_v + d_k
+        self._o_proj = nn.Parameter(torch.Tensor(num_heads * d_a, d_model))
+
+        self._d_model = d_model
+        self._d_k = d_k
+        self._d_v = d_v
+        self._d_a = d_a
+        self._num_heads = num_heads
+        self._memory_size = memory_size
+        self._relative_positional_encoding = relative_positional_encoding
+        self._add_positional_encoding = add_positional_encoding
+
+        self._attention_scale = 1. / d_k**0.5
+        self._mlp = torch.nn.Sequential(
+            FC(d_model, d_model, torch.relu_), FC(d_model, d_model))
+        self._norm1 = torch.nn.LayerNorm(d_model)
+        self._norm2 = torch.nn.LayerNorm(d_model)
+
+        l = 2 * memory_size - 1 if relative_positional_encoding else memory_size
+        self._positional_encoding = nn.Parameter(torch.Tensor(l, d_k))
+        # bias over query vectors when calculating score with keys. Introduced in [3].
+        self._qk_bias = nn.Parameter(torch.Tensor(1, num_heads, 1, d_k))
+        # bias over query vectors when calculating score with positional encodings.
+        # Introduced in [3].
+        self._qp_bias = nn.Parameter(torch.Tensor(1, num_heads, 1, d_k))
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        nn.init.xavier_uniform_(self._q_proj)
+        nn.init.xavier_uniform_(self._k_proj)
+        nn.init.xavier_uniform_(self._v_proj)
+        nn.init.xavier_uniform_(self._o_proj)
+        nn.init.zeros_(self._qk_bias)
+        nn.init.zeros_(self._qp_bias)
+        nn.init.uniform_(self._positional_encoding, -0.1, 0.1)
+        for l in self._mlp:
+            l.reset_parameters()
+
+    @staticmethod
+    def _shift(x, m):
+        """
+        y[i, j, :] <= x[n - 1 + i - j, :] for 0<=i<m, 0<=j<n
+        Args:
+            x: [2 * N - 1, d]
+        Returns:
+            [M, N, d]
+        """
+        n = (x.shape[0] + 1) // 2
+        # [M, N], index[i, j] = n - 1 + i - j
+        index = n - 1 + torch.arange(m).unsqueeze(-1) - torch.arange(n)
```

Comment:
In this relative positional encoding case, it seems that the query length should be checked so that it's no greater than the memory size.

Reply:
There is a check at line 1270.
You are right, the relative encoding only make sense if query and memory are derived from same sequence. query and memory can still be different though.